extension/observer/cfgardenobserver/extension.go

```diff
@@ -4,33 +4,337 @@
+	g := &cfGardenObserver{
+		config:     config,
+		logger:     logger,
+		once:       &sync.Once{},
+		containers: make(map[string]garden.ContainerInfo),
+		apps:       make(map[string]*resource.App),
+		cancel: func() {
+			// Safe value provided on initialisation
+		},
+	}
+	g.EndpointsWatcher = observer.NewEndpointsWatcher(g, config.RefreshInterval, logger)
 	return g, nil
 }
 
+func (g *cfGardenObserver) SyncApps() error {
+	g.containerMu.RLock()
+	containers := g.containers
+	g.containerMu.RUnlock()
+
+	g.appMu.Lock()
+	defer g.appMu.Unlock()
+	g.apps = make(map[string]*resource.App)
+	for _, info := range containers {
+		appID, err := infoToAppID(info)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := g.apps[appID]; ok {
+			continue
+		}
+
+		app, err := g.cf.Applications.Get(g.ctx, appID)
+		if err != nil {
+			return fmt.Errorf("error fetching application: %w", err)
+		}
+		g.apps[appID] = app
+	}
+
+	return nil
+}
+
+func (g *cfGardenObserver) App(info garden.ContainerInfo) (*resource.App, error) {
+	appID, err := infoToAppID(info)
+	if err != nil {
+		return nil, err
+	}
+
+	g.appMu.Lock()
+	defer g.appMu.Unlock()
+	app, ok := g.apps[appID]
+	if ok {
+		return app, nil
+	}
+
+	app, err = g.cf.Applications.Get(g.ctx, appID)
+	if err != nil {
+		return nil, err
+	}
+	g.apps[appID] = app
+
+	return app, nil
+}
+
+func (g *cfGardenObserver) Start(_ context.Context, _ component.Host) error {
+	gCtx, cancel := context.WithCancel(context.Background())
+	g.cancel = cancel
+	g.ctx = gCtx
+
+	g.garden = gardenClient.New(gardenConnection.New("unix", g.config.Garden.Endpoint))
+
+	var err error
+	g.cf, err = newCfClient(g.config.CloudFoundry)
+	if err != nil {
+		return err
+	}
+
+	if g.config.IncludeAppLabels {
+		g.once.Do(
+			func() {
+				go func() {
+					cacheRefreshTicker := time.NewTicker(g.config.CacheSyncInterval)
+					defer cacheRefreshTicker.Stop()
+
+					for {
+						select {
+						case <-g.ctx.Done():
+							return
+						case <-cacheRefreshTicker.C:
+							err = g.SyncApps()
+							if err != nil {
+								g.logger.Error("could not sync app cache", zap.Error(err))
+							}
+						}
+					}
+				}()
+			},
+		)
+	}
+
+	return nil
+}
+
+func (g *cfGardenObserver) Shutdown(_ context.Context) error {
+	g.cancel()
+	return nil
+}
+
 func (g *cfGardenObserver) ListEndpoints() []observer.Endpoint {
-	// TODO: Implement the logic to list the endpoints.
-	endpoints := make([]observer.Endpoint, 0)
+	var endpoints []observer.Endpoint
+
+	containers, err := g.garden.Containers(garden.Properties{})
+	if err != nil {
+		g.logger.Error("could not list containers", zap.Error(err))
+		return endpoints
+	}
+
+	infos := make(map[string]garden.ContainerInfo)
+	for _, c := range containers {
+		info, err := c.Info()
+		if err != nil {
+			g.logger.Error("error getting container info", zap.String("handle", c.Handle()), zap.Error(err))
+			continue
+		}
+
+		if info.State != "active" {
```

Comment:
Maybe use a central variable for "active". I see the cloud foundry golang package doesn't have one, so we can just declare our own.